README.md
# zfbzd

**zfbzd -> 支付宝账单**

获取指定时间段支付宝账单详情，可以根据时间区间进行筛选，可以过滤关键字，指定价格区间等

## Install

```bash
$ npm install zfbzd
```

## Usage

```js
const getBills = require('zfbzd')

;(async function() {
  let bills = await getBills({
    startDate: '2018-11-01',
    endDate: '2019-01-03',
  })
})()
```

## getBills(options)

必填项只有 `startDate` 和 `endDate`，用来指定账单时间区间，默认返回指定时间区间内全部记录

* `options` `{Object}`
  * `startDate` `{String}` 必填，账单区间开始时间，格式为 `yyyy-mm-dd` 
  * `endDate` `{String}` 必填，账单区间结束时间，格式为 `yyyy-mm-dd`
  * `puppeteerOptions` `{Object}` 选填，该值会被传递给 [zfbdl](https://github.com/funny-node/zfbdl#api)，详见其文档
  * `loginMethod` `{String}` 选填，模拟登录方式
    * `'scan'` 默认值，扫码登录
    * `'cookies'` cookies 登录，需要配置 `cookierc.js` 文件，详见 [这里](https://github.com/funny-node/zfbdl#loginbycookiesoptions)
    * `'pwd'` 账号密码登录
  * `username` `{String}` 选填，如果 `loginMethod` 配置了 `'pwd'`，则必须配置此项
  * `password` `{String}` 选填，如果 `loginMethod` 配置了 `'pwd'`，则必须配置此项
  * `turnPageDelay` `{Number}` 选填，翻页时间间隔，默认 `8000` 毫秒（其实默认的翻页时间间隔还得加上 `slowMo`，该值在 `puppeteerOptions` 中配置，默认 `500`）
  * `clickDatePickerDelay` `{Number}` 选填，模拟点击时间选择器的时间间隔，默认 `800` 毫秒（其实默认的点击时间间隔还得加上 `slowMo`，同上）
  * `billStartTime` `{String}` 选填，设置单天账单时间区间起始值，格式 `hh:mm`，默认值 `00:00`
  * `billEndTime` `{String}` 选填，设置单天账单时间区间结束值，格式 `hh:mm`，默认值 `24:00`
  * `minAmount` `{Number}` 选填，设置账单价格区间起始值，默认值 `Number.NEGATIVE_INFINITY`
  * `maxAmount` `{Number}` 选填，设置账单价格区间结束值，默认值 `Number.POSITIVE_INFINITY`
  * `billExcludes` `{Array}` 选填，过滤账单关键字
  * `sorting` `{String}` 选填，设置返回账单排序规则
    * `'desc'` 默认值，倒序返回
    * `'asc'` 正序返回
* return: `<Promise<Array<Object>>>`
  * `date` `{String}` 该条账单日期，格式 `yyyy-mm-dd` 
  * `time` `{String}` 该条账单时间，格式 `hh:mm`
  * `year` `{Number}` 年
  * `month` `{Number}` 月
  * `day` `{Number}` 日
  * `timestamp` `{Number}` 该条消费记录时间戳
  * `amoumt` `{Number}` 该条账单具体账目
  * `outHTML` `{String}` 该条账单 HTML 字符串

## License

MIT



CHANGELOG.md
# CHANGELOG

* 1.0.1 - 将账单返回中的 `今天` `昨天` 替换成具体日期
* 1.1.0 - 增加可配置参数 `sorting` 

util/helper.js
/**
 * 格式化时间，月、日加前导 0
 * @param {String} dateStr 
 * @return {String}
 */
function formateDate(dateStr) {
  let [year, month, day] = dateStr.split('-')
  return [year, String(month).padStart(2, '0'), String(day).padStart(2, '0')].join('.')
}

/**
 * 将数据根据日期升序或者降序排列
 * @param {Array} ret 
 * @param {String} sorting 
 */
function reorder(ret, sorting) {
  if (!sorting) {
    return ret
  } else if (sorting === 'asc') { // 升序
    ret.sort((a, b) => {
      if (a.year !== b.year) return a.year - b.year 
      else if (a.month !== b.month) return a.month - b.month 
      else return a.day - b.day
    })
  } else if (sorting === 'desc') { // 降序
    ret.sort((a, b) => {
      if (a.year !== b.year) return b.year - a.year 
      else if (a.month !== b.month) return b.month - a.month 
      else return b.day - a.day
    })
  } else {
    throw new Error('参数错误')
  }

  return ret
}

module.exports = {
  formateDate,
  reorder,
}

lib/getDetailInPage.js
/**
 * 解析 DOM, 获取该页面的账单数据 
 * @https://consumeprod.alipay.com/record/standard.htm
 * @param {Object} options 
 * @return {Array<Object>}
 */
function getDetailInPage(options) {
  let {
    billStartTime = '00:00',  // 账单时间区间
    billEndTime = '24:00',
    minAmount = Number.NEGATIVE_INFINITY, // 账单价格区间
    maxAmount = Number.POSITIVE_INFINITY,
    billExcludes = [], // 根据关键字过滤某些账单
  } = options

  const ret = []
  const items = document.querySelectorAll('.J-item')

  items.forEach(item => {
    let date = item.querySelector('.time p').innerHTML.trim()
    const time = item.querySelector('.time .text-muted').innerHTML.trim()
    const amount = +item.querySelector('.amount .amount-pay').innerHTML.trim().replace(/ /g, '')
    const outerHTML = item.outerHTML 

    // 根据账单时间区间进行删选
    // 不在区间内，则放弃这条数据
    if (!isTimeInTheInterval(time, billStartTime, billEndTime)) return 

    // 根据 billExcludes 进行删选
    // 排除一些不必要的账单信息
    if (isIncludeExcludes(outerHTML, billExcludes)) return 

    // 价格不在区间里
    if (amount < minAmount || amount > maxAmount) return 
    
    if (date === '今天') date = changeTimestampToFormate(+new Date)
    else if (date === '昨天') date = changeTimestampToFormate(+new date - 3600*24*1000)
    
    let [year, month, day] = date.split('-').map(item => +item)

    ret.push({
      date, 
      time, 
      year,
      month, 
      day,
      timestamp: +new Date(`${date} ${time}`),
      amount,
      outerHTML,
    })
  })

  return ret
  
  /** 
  * 经测试
  * page.evaluateHandle(pageFunction[, ...args])
  * pageFunction 方法里调用的方法不能是在外面的，所以只能写在 getDetailInPage 方法里面 ..
  */ 
  
  /**
   * 账单内容是否包含了需要排除的词
   * @param {String} outerHTML 
   * @param {Array} billExcludes 
   * @return {Boolean}
   */
  function isIncludeExcludes(outerHTML, billExcludes) {
    return billExcludes.some(item => outerHTML.includes(item))
  }

  /**
   * 判断一个时间是否在区间内
   * @param {String} time `hh:ss` 格式
   * @param {String} startTime `hh:ss` 格式
   * @param {String} endTime `hh:ss` 格式
   * @return {Boolean}
   */
  function isTimeInTheInterval(time, startTime, endTime) {
    let times = changeTimeToSeconds(time)
    let startTimes = changeTimeToSeconds(startTime)
    let endTimes = changeTimeToSeconds(endTime)
    
    return times >= startTimes && times <= endTimes
  }

  /**
   * 将 `hh:ss` 格式的时间格式转为秒
   * @param {String} time 
   * @return {Number}
   */
  function changeTimeToSeconds(time) {
    let [hour, second] = time.split(':').map(item => +item)
    return hour * 60 + second
  }

  /**
   * 将时间戳转为 'yyyy-mm-dd` 格式
   * @param {Number} timestamp 
   */
  function changeTimestampToFormate(timestamp) {
    const d = new Date(timestamp)
    const year = d.getFullYear()
    const month = d.getMonth() + 1 
    const day = d.getDate()
    return [year, String(month).padStart(2, '0'), String(day).padStart(2, '0')].join('-')
  }
}

module.exports = getDetailInPage


lib/setIntervalDate.js
/**
 * 改变账单起始以及截止日期
 * @param {Page} page 
 * @param {String} startDate 账单起始日期
 * @param {String} endDate 账单结束日期
 * @param {Number} clickDatePickerDelay 模拟点击时间选择器的间隔时间，单位毫秒
 */
function setIntervalDate(page, startDate, endDate, clickDatePickerDelay) {
  return new Promise(async resolve => {
    /** 
    * 支付宝账单页默认删选区间是最近一个月
    * 比如今天是 2019-02-16
    * 那么默认显示的删选日期是 2019-01-16 ~ 2019-02-16
    * 我们需要手动将其置为 [startDate, endDate]
    */
    
    // 改变开始日期
    // 通过计算获取默认的开始时间
    let cur_date = new Date()
    let start_year = cur_date.getFullYear()
    let start_month = cur_date.getMonth()
    // 目标时间
    let target_start_year = +startDate.split('.')[0]
    let target_start_month = +startDate.split('.')[1]
    // 相隔的月数，也是需要点击点一个月按钮的次数
    let times = (start_year * 12 + start_month) - (target_start_year * 12 + target_start_month)

    await page.waitFor(clickDatePickerDelay)
    await page.click('#beginDate')
    await page.waitFor(clickDatePickerDelay)

    // 这里有个特殊情况，可能会是 -1，需要特判
    if (times === -1) {
      await page.click('body > div.jui-calendar > div.jui-calendar-pannel > span.jui-calendar-control.jui-calendar-next-month')
      await page.waitFor(clickDatePickerDelay)
    } else {
      while (times --) {
        await page.click('body > div.jui-calendar > div.jui-calendar-pannel > span.jui-calendar-control.jui-calendar-prev-month')
        await page.waitFor(clickDatePickerDelay)
      }
    }
    
    await page.click(`body > div.jui-calendar > div.jui-calendar-container > table.jui-calendar-date > tbody [data-value="${startDate}"]`)
    
    // 改变结束日期
    // 默认的结束日期就是今天
    let end_date = new Date()
    let end_year = end_date.getFullYear()
    let end_month = end_date.getMonth() + 1
    // 目标时间
    let target_end_year = +endDate.split('.')[0]
    let target_end_month = +endDate.split('.')[1]
    times = (end_year * 12 + end_month) - (target_end_year * 12 + target_end_month)
    
    await page.waitFor(clickDatePickerDelay)
    await page.click('#endDate')
    await page.waitFor(clickDatePickerDelay)

    while(times --) {
      await page.click('body > div.jui-calendar > div.jui-calendar-pannel > span.jui-calendar-control.jui-calendar-prev-month')
      await page.waitFor(clickDatePickerDelay)
    }

    await page.click(`body > div.jui-calendar > div.jui-calendar-container > table.jui-calendar-date > tbody [data-value="${endDate}"]`)
    
    resolve()
  })
}

module.exports = setIntervalDate


index.js
const { loginByScanQR, loginByCookies, loginByPwd } = require('zfbdl')
const { TimeoutError } = require('puppeteer/Errors')
const { formateDate, reorder } = require('./util/helper')
const setIntervalDate = require('./lib/setIntervalDate')
const getDetailInPage = require('./lib/getDetailInPage')

/**
 * 根据时间区间获取账单详情
 * @param {Object} options 必填
 * @return {Array<Object>}
 */
function getBills(options = {}) {
  let {
    startDate, // 必填
    endDate, // 必填
    loginMethod = 'scan', // 默认扫码登录
    username, // 如果 loginMethod === 'pwd' 则必填
    password, // 如果 loginMethod === 'pwd' 则必填
    puppeteerOptions = {},
    turnPageDelay = 8000, // 默认翻页之间停顿 8s
    clickDatePickerDelay = 800, // 默认时间选择器操作之间停顿 0.8s
    sorting,
  } = options

  startDate = formateDate(startDate)
  endDate = formateDate(endDate)

  return new Promise(async resolve => {
    let page, browser, loginRetData

    // 模拟登录
    if (loginMethod === 'scan') {
      loginRetData = await loginByScanQR(puppeteerOptions)
    } else if (loginMethod === 'cookies') {
      loginRetData = await loginByCookies(puppeteerOptions)
    } else if (loginMethod === 'pwd') {
      loginRetData = await loginByPwd(username, password, puppeteerOptions)
    }
    
    page = loginRetData.page 
    browser = loginRetData.browser
    
    await page.goto('https://consumeprod.alipay.com/record/standard.htm')

    // 如果操作频繁，可能需要扫码验证
    try {
      await page.waitForSelector('.J-item', { timeout: 0 })
    } catch (e) {
      if (e instanceof TimeoutError) {
        console.log('长时间没有扫描验证码，被当作机器啦！')
      }
    }

    // 设置时间删选区间
    await setIntervalDate(page, startDate, endDate, clickDatePickerDelay)
    
    // 账单数据
    let cost_data = []

    while (true) {
      // 分析页面数据
      let ret = await page.evaluate(getDetailInPage, options)
      cost_data.push(...ret)

      // 已经到了最后一页
      if (!await page.$('#J_home-record-container > div.amount-bottom > div > div.fn-clear.action-other.action-other-show > div.page.fn-right > div > a.page-next.page-trigger')) {
        break
      }

      // 每次翻页前等待
      await page.waitFor(turnPageDelay)
      
      await Promise.all([
        page.click('#J_home-record-container > div.amount-bottom > div > div.fn-clear.action-other.action-other-show > div.page.fn-right > div > a.page-next.page-trigger'),
        page.waitForNavigation({ waitUntil: 'networkidle2' })
      ])

      // 如果操作频繁，可能需要扫码验证
      try {
        await page.waitForSelector('.J-item', { timeout: 0 })
      } catch (e) {
        if (e instanceof TimeoutError) {
          console.log('长时间没有扫描验证码，被当作机器啦！')
        }
      }
    }

    resolve(reorder(cost_data, sorting))

    // 这样才能关闭 node 进程
    await browser.close() 
  })
}

module.exports = getBills
